src/types/type_param.rs

```diff
@@ -19,55 +20,84 @@ use super::{ClassicType, SimpleType};
 #[derive(Clone, Debug, PartialEq, Eq, serde::Deserialize, serde::Serialize)]
 #[non_exhaustive]
 pub enum TypeArg {
-    /// Where the TypeDef declares that an argument is a [TypeParam::Type]
+    /// Where the (Type/Op)Def declares that an argument is a [TypeParam::Type]
     Type(SimpleType),
-    /// Where the TypeDef declares that an argument is a [TypeParam::ClassicType],
+    /// Where the (Type/Op)Def declares that an argument is a [TypeParam::ClassicType],
     /// it'll get one of these (rather than embedding inside a Type)
     ClassicType(ClassicType),
-    /// Where the TypeDef declares a [TypeParam::Int]
+    /// Where the (Type/Op)Def declares that an argument is a [TypeParam::HashableType],
+    /// this is the value.
+    HashableType(HashableType),
+    /// Where the (Type/Op)Def declares a [TypeParam::Value] of type [HashableType::Int], a constant value thereof
     Int(HugrIntValueStore),
-    /// Where an argument has type [TypeParam::List]`<T>` - all elements will implicitly
-    /// be of the same variety of TypeArg, representing a `T`.
+    /// Where the (Type/Op)Def declares a [TypeParam::Value] of type [HashableType::String], here it is
+    String(String),
+    /// Where the (Type/Op)Def declares a [TypeParam::List]`<T>` - all elements will implicitly
+    /// be of the same variety of TypeArg, i.e. `T`s.
     List(Vec<TypeArg>),
-    /// Where the TypeDef declares a [TypeParam::Value]
-    Value(serde_yaml::Value),
+    /// Where the TypeDef declares a [TypeParam::Value] of [Container::Opaque]
+    CustomValue(serde_yaml::Value),
 }
 
 /// Checks a [TypeArg] is as expected for a [TypeParam]
 pub fn check_type_arg(arg: &TypeArg, param: &TypeParam) -> Result<(), TypeArgError> {
     match (arg, param) {
-        (TypeArg::Type(_), TypeParam::Type) => (),
-        (TypeArg::ClassicType(_), TypeParam::ClassicType) => (),
-        (TypeArg::Int(_), TypeParam::Int) => (),
+        (TypeArg::Type(_), TypeParam::Type) => Ok(()),
+        (TypeArg::ClassicType(_), TypeParam::ClassicType) => Ok(()),
+        (TypeArg::HashableType(_), TypeParam::HashableType) => Ok(()),
         (TypeArg::List(items), TypeParam::List(ty)) => {
             for item in items {
                 check_type_arg(item, ty.as_ref())?;
             }
+            Ok(())
         }
-        (TypeArg::Value(_), TypeParam::Value) => (),
-        _ => {
-            return Err(TypeArgError::TypeMismatch(arg.clone(), param.clone()));
+        (TypeArg::Int(v), TypeParam::Value(HashableType::Int(width))) => {
+            check_int_fits_in_width(*v, *width).map_err(TypeArgError::Int)
         }
-    };
-    Ok(())
+        (TypeArg::String(_), TypeParam::Value(HashableType::String)) => Ok(()),
+        (arg, TypeParam::Value(HashableType::Container(ctr))) => match ctr {
+            Container::Opaque(_) => match arg {
+                TypeArg::CustomValue(_) => Ok(()), // Are there more checks we should do here?
```

Comment:
I don't think there's anything yet, but we might want to allow `TypeDef` providers to give code that checks a value is correct (e.g. deserializes to something appropriate)